text/3485-feature-documentation.md

```diff
@@ -0,0 +1,195 @@
+- Feature Name: feature-documentation
+- Start Date: 2023-09-09
+- RFC PR: [rust-lang/rfcs#3485](https://github.com/rust-lang/rfcs/pull/3485)
+- Rust Issue:
+  [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+[summary]: #summary
+
+This RFC describes a new key to under `features` in `Cargo.toml` for
+documentation. This will allow Cargo to display this information to the user and
+provide a way for `rustdoc` to eventually render this data (how this is rendered
+is outside the scope of this RFC).
+
+Please see the parent meta RFC for background information: [`feature-metadata`].
+
+# Motivation
+
+[motivation]: #motivation
+
+Cargo features have become extremely widely used, with many crates having at
+least some level of configuration and larger crates winding up with tens of
+gates. Desipte being a first class component of crate structure, they suffer
+from a documentation problem: users need to maintain documentation separate from
+feature definition, typically a manually-created table within API docs.
+
+This RFC proposes adding feature documentation to `Cargo.toml`, which will allow
+for keeping feature definitions and documentation together.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+A new `doc` key will be allowed within a feature's table. This key provides a
+markdown docstring describing the feature. Like with `#[doc(...)]`, the first
+line will be treated as a summary.
+
+```toml
+[features]
+# Feature without documentation
+foo = []
+
+# Short documentation comment
+bar = { enables = ["foo"], doc = "simple docstring here"}
+
+# Tables are preferred for longer descriptions
+[features.corge]
+enables = ["bar", "baz"]
+doc = """
```

Comment:
Another good question raised was "should we also support intra-doc links in this documentation?". I personally think we should and make the context the same as the crate top-level. What do you think?

Reply:
Ah I see you mentioned it below, my bad.

Reply:
Thanks for the response!

The follow-up question from me would be: If there any compatibility issues if we hadn't done this and rustdoc change all together? If not then this RFC can safely go first.

Reply:
I'm not sure. From cargo perspective, whether there are intra-doc links or not in this documentation doesn't matter. But it'll definitely need to be mentioned when support will be discussed in rustdoc.

For the current case, I think `cargo` should mention that intra-doc links may be supported when rustdoc support this option and that's it. What do you think?

Reply:
Sounds good to me.

@tgross35 could you add something like this? (or whichever way you'd like to rephase this)


```patch
- Rustdoc can build on this to show feature documentation.
+ Rustdoc can build on this to show feature documentation.
+ If this RFC gets stabilized before any corresponding change in rustdoc,
+ its documentation should highlight that rustdoc may parse the description and support intra-doc links in the feature.
+ Users need to be aware of this potential incompatibility.
```

Reply:
Added, thank you!